gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,432 @@
+        st.insert([t], filtration=results[t])
+    for i in range(connections.shape[0]):
+        in1 = connections[i, 0]
+        in2 = connections[i, 1]
+        fil = max(results[in1], results[in2])
+        st.insert([in1, in2], filtration=fil)
+    st.expansion(expansion)
+    st.compute_persistence()
+    tmp = st.persistence_pairs()
+    persistences = np.zeros([len(tmp), 2])
+    pers_indices = np.zeros([len(tmp), 2])
+    odd_one_out = 0
+    for i in range(len(tmp)):
+        inds1 = tmp[i][0]
+        inds2 = tmp[i][1]
+        if len(inds2) == 0:
+            odd_one_out = i
+            continue
+        for index in range(len(inds1)):
+            value = results[inds1[index]]
+            if value > persistences[i, 0]:
+                persistences[i, 0] = value
+                pers_indices[i, 0] = inds1[index]
+        for index in range(len(inds2)):
+            value = results[inds2[index]]
+            if value > persistences[i, 1]:
+                persistences[i, 1] = value
+                pers_indices[i, 1] = inds2[index]
+    out_pers = np.delete(persistences, odd_one_out, axis=0)
+    out_pers_ind = np.delete(pers_indices, odd_one_out, axis=0)
+    persts = out_pers_ind.astype(int)
+    return (out_pers, persts)
+
+
+def compute_critical_points(
+    net: torch.nn.Module,
+    coords: Tensor,
+    dummy_loader: DataLoader[Tuple[Tensor, ...]],
+    connections: np.ndarray,
+    expansion: int,
+    ind: int = 0,
+) -> Tensor:
+    """
+    A helper file for topological regularizer
+    Given a neural network and a dataset, computes the
+    ephemeral critical values and points of the linear
+    Morse function defined by the dataset and connections.
+    The filtering directions is specified by ind (0: Positive, 1 negative)
+    A helper file for topological regularizer
+    Args:
+        model: torch model
+        coords: torch.tensor, the nodes of the linear surrogate function
+        dummy_loader: torch dataloader containing the coords.
+            connections: numpy_array. An n by 2 numpy array describing
+            pairwise connections between the nodes.
+        expansion: The size of the clique expansion
+        ind: integer (for now boolean) describing filtration direction
+            0: positive filtration, 1: negative filtration
+    Returns
+    -------
+        coordinates: The coordinates of the critical points that
+        contribute to the complexity of the decision boundary.
+    """
+    res = evaluate_model_on_grid(dummy_loader, net)
+    if ind == 1:
+        res = 1 - res
+    out_pers, persts = get_persistent_pairs(res, connections, expansion)
+
+    """ Get only the persistent pairs that contribute to
+    the complexity of the decision boundary:
+    These are the ones that are on the different side of the boundary
+    """
+    tmp = out_pers - 0.5
+    cuts = (tmp[:, 0] * tmp[:, 1]) < 0
+
+    goodinds = np.argmin(abs(tmp[cuts, :]), 1)
+    tmp_inds = persts[cuts, :]
```

Comment:
also here the naming is not consistent `goodinds` without `_ ` and `tmp_inds` with `_`.

Reply:
Also `good` is not very descriptive.

Reply:
done